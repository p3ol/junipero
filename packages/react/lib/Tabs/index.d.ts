import { ReactNode, ComponentPropsWithRef, MutableRefObject } from 'react';

export declare type TabsRef = {
  activeTab: number;
  isJunipero: boolean;
  innerRef: MutableRefObject<any>;
};

declare interface TabsProps extends ComponentPropsWithRef<any> {
  active?: number;
  children?: ReactNode | JSX.Element;
  className?: string;
  disabled?: boolean;
  filterTab?(child: ReactNode | JSX.Element): boolean;
  onToggle?(index: number): void;
  ref?: MutableRefObject<TabsRef | undefined>;
}

declare function Tabs(props: TabsProps): ReactNode | JSX.Element;

export default Tabs;
