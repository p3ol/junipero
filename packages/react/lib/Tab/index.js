import { forwardRef, useRef, useImperativeHandle } from 'react';
import PropTypes from 'prop-types';
import { classNames } from '@junipero/core';

const Tab = forwardRef(({
  className,
  tag: Tag = 'div',
  title: _,
  ...rest
}, ref) => {
  const innerRef = useRef();

  useImperativeHandle(ref, () => ({
    innerRef,
    isJunipero: true,
  }));

  return (
    <Tag
      { ...rest }
      className={classNames(
        'tab',
        className,
      )}
      ref={ref}
    />
  );
});

Tab.displayName = 'Tab';
Tab.propTypes = {
  tag: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.node,
    PropTypes.object,
    PropTypes.func,
  ]),
  title: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.node,
    PropTypes.func,
  ]),
};

export default Tab;
