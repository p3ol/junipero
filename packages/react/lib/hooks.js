import { useContext } from 'react';

import { FieldControlContext } from './contexts';

export const useFieldControl = () => useContext(FieldControlContext);
