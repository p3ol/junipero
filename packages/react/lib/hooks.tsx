import { useContext } from 'react';

import {
  AlertsContextType,
  AlertsContext,
  DropdownContext,
  DropdownContextType,
  FieldContextType,
  FieldControlContext,
  ListContext,
  ListContextType,
  ToastsContext,
  ToastsContextType,
  ModalContext,
  ModalContextType,
} from './contexts';

export const useFieldControl =
  () => useContext<FieldContextType>(FieldControlContext);
export const useDropdown = () =>
  useContext<DropdownContextType>(DropdownContext);
export const useAlerts = () => useContext<AlertsContextType>(AlertsContext);
export const useList = () => useContext<ListContextType>(ListContext);
export const useToasts = () => useContext<ToastsContextType>(ToastsContext);
export const useModal = () => useContext<ModalContextType>(ModalContext);

export { AlertsContext, DropdownContext, FieldControlContext, ListContext, ToastsContext, ModalContext };
