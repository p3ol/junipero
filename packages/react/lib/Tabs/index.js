import {
  Children,
  forwardRef,
  useEffect,
  useImperativeHandle,
  useRef,
  useState,
  useMemo,
} from 'react';
import PropTypes from 'prop-types';
import { classNames } from '@junipero/core';

import Tab from '../Tab';

const Tabs = forwardRef(({
  className,
  children,
  active,
  tabs,
  disabled = false,
  filterTab = child => child.type === Tab,
  onToggle,
  ...rest
}, ref) => {
  const innerRef = useRef();
  const [activeTab, setActiveTab] = useState(active);

  useImperativeHandle(ref, () => ({
    tabs,
    activeTab,
    isJunipero: true,
    innerRef,
  }));

  useEffect(() => {
    setActiveTab(active ?? 0);
  }, [active]);

  const onClick_ = (tab, index, e) => {
    e?.preventDefault?.();

    if (disabled || tab.props.disabled) {
      return;
    }

    setActiveTab(index);
    onToggle?.(index);
  };

  const availableTabs = useMemo(() => (
    tabs
      ? tabs.map((t, i) => <Tab key={i} title={t.title}>{ t.content }</Tab>)
      : Children.toArray(children).filter(filterTab)
  ), [tabs, children]);

  return (
    <div
      { ...rest }
      ref={innerRef}
      className={classNames(
        'junipero',
        'tabs',
        className,
      )}
    >
      <ul className="titles">
        { availableTabs.map((tab, index) => (
          <li
            key={index}
            className={classNames(
              'title',
              {
                active: index === activeTab,
                disabled: tab.props.disabled,
              }
            )}
          >
            <a href="#" onClick={onClick_.bind(null, tab, index)}>
              { tab.props.title }
            </a>
          </li>
        ))}
      </ul>

      <div className="content">
        { availableTabs[activeTab] }
      </div>
    </div>
  );
});

Tabs.displayName = 'Tabs';
Tabs.propTypes = {
  active: PropTypes.number,
  tabs: PropTypes.arrayOf(PropTypes.shape({
    title: PropTypes.any,
    content: PropTypes.any,
  })),
  disabled: PropTypes.bool,
  onToggle: PropTypes.func,
  filterTab: PropTypes.func,
};

export default Tabs;
