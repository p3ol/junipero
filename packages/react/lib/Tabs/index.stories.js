import { action } from '@storybook/addon-actions';

import Tabs from './';
import Tab from '../Tab';

export default { title: 'react/Tabs' };

export const basic = () => (
  <Tabs onToggle={action('change')}>
    <Tab title="Title One">Content One</Tab>
    <Tab title="Title Two">Content Two</Tab>
  </Tabs>
);
