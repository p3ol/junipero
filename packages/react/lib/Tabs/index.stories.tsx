import { action } from 'storybook/actions';

import Tabs from '.';
import Tab from '../Tab';

export default { title: 'react/Tabs' };

export const Basic = () => (
  <Tabs onToggle={action('change')}>
    <Tab title="Title One">Content One</Tab>
    <Tab title="Title Two">Content Two</Tab>
  </Tabs>
);

export const WithTabsProp = () => (
  <Tabs
    tabs={[
      { title: 'Title One', content: 'Content One' },
      { title: 'Title Two', content: 'Content Two' },
    ]}
    onToggle={action('change')}
  />
);
