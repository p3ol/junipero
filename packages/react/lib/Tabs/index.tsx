import {
  type RefObject,
  type ReactNode,
  type MouseEvent,
  type ReactElement,
  type ComponentPropsWithoutRef,
  Children,
  useEffect,
  useImperativeHandle,
  useRef,
  useState,
  useMemo,
} from 'react';
import { classNames } from '@junipero/core';

import type { JuniperoRef, SpecialComponentPropsWithRef } from '../types';
import Tab, { type TabProps, type TabObject } from '../Tab';

export declare interface TabsRef extends JuniperoRef {
  activeTab: number;
  tabs: TabObject[];
  innerRef: RefObject<HTMLDivElement>;
}

export declare interface TabsProps extends Omit<
  SpecialComponentPropsWithRef<'div', TabsRef>,
  'onToggle'
> {
  active?: number;
  disabled?: boolean;
  tabs?: TabObject[];
  filterTab?(child: ReactElement<ComponentPropsWithoutRef<any>>): boolean;
  onToggle?(index: number): void;
}

const Tabs = ({
  ref,
  className,
  children,
  active,
  tabs,
  disabled = false,
  filterTab = (child: ReactElement<ComponentPropsWithoutRef<any>>) =>
    typeof child !== 'string' && child.type === Tab,
  onToggle,
  ...rest
}: TabsProps) => {
  const innerRef = useRef<HTMLDivElement>(null);
  const [activeTab, setActiveTab] = useState(active);

  useImperativeHandle(ref, () => ({
    tabs,
    activeTab,
    isJunipero: true,
    innerRef,
  }));

  useEffect(() => {
    setActiveTab(active ?? 0);
  }, [active]);

  const onClick_ = (
    tab: TabObject,
    index: number,
    e: MouseEvent<HTMLAnchorElement>
  ) => {
    e?.preventDefault?.();

    if (disabled || tab.props.disabled) {
      return;
    }

    setActiveTab(index);
    onToggle?.(index);
  };

  const availableTabs = useMemo<ReactNode[]>(() => (
    tabs
      ? tabs.map((t, i) => <Tab key={i} title={t.title}>{ t.content }</Tab>)
      : Children.toArray(children).filter(filterTab)
  ), [tabs, children, filterTab]);

  return (
    <div
      { ...rest }
      ref={innerRef}
      className={classNames(
        'junipero',
        'tabs',
        className,
      )}
    >
      <ul className="titles">
        { availableTabs.map((tab: ReactElement<TabProps>, index: number) => (
          <li
            key={index}
            className={classNames(
              'title',
              {
                active: index === activeTab,
                disabled: tab.props?.disabled,
              }
            )}
          >
            <a href="#" onClick={onClick_.bind(null, tab, index)}>
              { tab.props?.title }
            </a>
          </li>
        )) }
      </ul>

      <div className="content">
        { availableTabs[activeTab] }
      </div>
    </div>
  );
};

Tabs.displayName = 'Tabs';

export default Tabs;
