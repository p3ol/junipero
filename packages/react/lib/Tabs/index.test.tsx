import { render, fireEvent } from '@testing-library/react';
import { vi } from 'vitest';

import Tabs from '.';
import Tab from '../Tab';

describe('<Tabs />', () => {
  it('should render', () => {
    const { container, unmount } = render(
      <Tabs><Tab title="Tab 1">Content</Tab></Tabs>
    );
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should fire onToggle event when changing tab', () => {
    const onToggle = vi.fn();
    const { container, getByText, unmount } = render(
      <Tabs onToggle={onToggle}>
        <Tab title="Tab 1">One</Tab>
        <Tab title="Tab 2">Two</Tab>
      </Tabs>
    );
    fireEvent.click(getByText('Tab 2'));
    expect(onToggle).toHaveBeenCalledWith(1);
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('shouldn\'t change active tab if tab is disabled', () => {
    const onToggle = vi.fn();
    const { container, getByText, unmount } = render(
      <Tabs onToggle={onToggle}>
        <Tab title="Tab 1">One</Tab>
        <Tab title="Tab 2" disabled={true}>Two</Tab>
      </Tabs>
    );
    fireEvent.click(getByText('Tab 2'));
    expect(container).toMatchSnapshot();
    unmount();
  });

  it('should allow to render tabs from prop', () => {
    const onToggle = vi.fn();
    const { container, getByText, unmount } = render(
      <Tabs
        tabs={[
          { title: 'Tab 1', content: 'One' },
          { title: 'Tab 2', content: 'Two' },
        ]}
        onToggle={onToggle}
      />
    );
    expect(container).toMatchSnapshot('First tab');
    fireEvent.click(getByText('Tab 2'));
    expect(onToggle).toHaveBeenCalledWith(1);
    expect(container).toMatchSnapshot('Second tab');
    unmount();
  });
});
