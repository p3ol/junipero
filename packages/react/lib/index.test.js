import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import TextField from './TextField';

describe('mockState: inside react', () => {
  it('should fail update the state correctly and return the old ' +
    'state', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch({ value: { ...state.value, content: ['foo'] } });
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch({ value: { ...state.value, text: value } });
      };

      const onContentChange = () => {
        dispatch({
          value: { ...state.value, content: [...state.value.content, 'bar'] },
        });
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content but with old text value');

    unmount();
  });

  it('should not fail update the state correctly and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
