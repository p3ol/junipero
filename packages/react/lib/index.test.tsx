import { useReducer } from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { mockState } from '@junipero/core';
import { useTimeout } from '@junipero/hooks';

import type { FieldContent } from './types';
import TextField from './TextField';

interface State {
  value: {
    text: string;
    content: string[];
  };
}

describe('mockState: inside react', () => {
  it('should correctly update the state and return the new ' +
    'state when using mockState with a callback', async () => {
    const Comp = () => {
      const [state, dispatch] = useReducer(mockState<State>, {
        value: {
          text: 'This is a test',
          content: [],
        },
      });

      useTimeout(() => {
        dispatch(s => ({ value: { ...s.value, content: ['foo'] } }));
      }, 1, []);

      const onTextChange = ({ value }: FieldContent) => {
        dispatch(s => ({ value: { ...s.value, text: value } }));
      };

      const onContentChange = () => {
        dispatch(s => ({
          value: { ...s.value, content: [...s.value.content, 'bar'] },
        }));
      };

      return (
        <div>
          <TextField
            data-testid="Field"
            value={state.value.text}
            onChange={onTextChange}
          />
          <button onClick={onContentChange}>Add content</button>
          <div>Text: { state.value.text }</div>
          <div>Content: { state.value.content.join(', ') }</div>
        </div>
      );
    };

    const { container, unmount } = render(<Comp />);

    fireEvent.change(screen.getByTestId('Field'),
      { target: { value: 'This is a test 2' } });

    await waitFor(() => screen.getByText('Content: foo'));
    expect(container).toMatchSnapshot('With updated text field value');

    fireEvent.click(screen.getByText('Add content'));

    expect(container)
      .toMatchSnapshot('With updated content AND updated text field valud');

    unmount();
  });
});
