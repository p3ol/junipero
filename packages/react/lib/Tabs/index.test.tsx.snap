// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> should allow to render tabs from prop: First tab 1`] = `
<div>
  <div
    class="junipero tabs"
  >
    <ul
      class="titles"
    >
      <li
        class="title active"
      >
        <a
          href="#"
        />
      </li>
      <li
        class="title"
      >
        <a
          href="#"
        />
      </li>
    </ul>
    <div
      class="content"
    >
      <div
        class="tab"
      >
        One
      </div>
    </div>
  </div>
</div>
`;

exports[`<Tabs /> should render 1`] = `
<div>
  <div
    class="junipero tabs"
  >
    <ul
      class="titles"
    >
      <li
        class="title active"
      >
        <a
          href="#"
        />
      </li>
    </ul>
    <div
      class="content"
    >
      <div
        class="tab"
      >
        Content
      </div>
    </div>
  </div>
</div>
`;
